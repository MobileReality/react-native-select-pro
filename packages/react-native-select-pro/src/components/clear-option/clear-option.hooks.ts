import { useSelectContext } from '../../context';

export const useClearOption = () => {
    const {
        styles: mainStyles,
        disabled,
        clearOptionButtonProps,
        clearOptionImageProps,
        dispatch,
        searchValue,
        selectedOption,
        selectedOptionIndex,
        onRemove,
    } = useSelectContext();

    const { clear } = mainStyles?.select ?? {};
    const { icon: iconCustomStyles, container: containerCustomStyles } = clear ?? {};

    const removeSingleOption = () => {
        dispatch({
            type: 'selectOption',
            payload: {
                selectedOption: null,
                selectedOptionIndex: -1,
            },
        });

        const isSearchable = typeof searchValue === 'string';
        if (isSearchable) {
            dispatch({
                type: 'setSearchValue',
                payload: '',
            });
        }
    };
    const onPressClearOption = () => {
        if (disabled) {
            return;
        }
        removeSingleOption();
        if (onRemove && selectedOption) {
            // callback
            onRemove(selectedOption, selectedOptionIndex);
        }
    };

    return {
        disabled,
        clearOptionButtonProps,
        clearOptionImageProps,
        iconCustomStyles,
        containerCustomStyles,
        onPressClearOption,
    };
};
