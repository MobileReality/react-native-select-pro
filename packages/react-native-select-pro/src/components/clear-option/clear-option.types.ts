import type { OnPressRemove, OptionalToRequired } from '../../types';

export type ClearOptionProps = OptionalToRequired<{
    onPressRemove: OnPressRemove;
}>;
