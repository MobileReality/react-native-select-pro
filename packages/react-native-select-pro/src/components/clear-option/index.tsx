import React, { ComponentProps } from 'react';
import {
    Image,
    ImageStyle,
    Pressable,
    StyleSheet,
    ViewStyle,
} from 'react-native';

import type { OptionalToRequired } from '../../helpers/types/OptionalToRequired';
import type { SelectControl } from '../select-control';

type FromSelectControlProps = Pick<
    ComponentProps<typeof SelectControl>,
    | 'selectControlClearOptionA11yLabel'
    | 'disabled'
    | 'selectControlClearOptionButtonHitSlop'
    | 'selectControlClearOptionButtonStyle'
    | 'selectControlClearOptionImageStyle'
> & {
    onPressRemove: () => void;
};

type ClearOptionProps = OptionalToRequired<FromSelectControlProps>;

export const ClearOption = ({
    selectControlClearOptionA11yLabel,
    disabled,
    selectControlClearOptionButtonHitSlop,
    selectControlClearOptionButtonStyle,
    selectControlClearOptionImageStyle,
    onPressRemove,
}: ClearOptionProps) => {
    return (
        <Pressable
            accessibilityLabel={
                selectControlClearOptionA11yLabel || 'Clear a chosen option'
            }
            accessibilityRole="button"
            accessible={true}
            disabled={disabled}
            hitSlop={
                selectControlClearOptionButtonHitSlop
                    ? selectControlClearOptionButtonHitSlop
                    : { right: 3, left: 3 }
            }
            style={[styles.xIconWrapper, selectControlClearOptionButtonStyle]}
            onPress={onPressRemove}
        >
            <Image
                source={require('./../../assets/icons/x.png')}
                style={[styles.xIcon, selectControlClearOptionImageStyle]}
            />
        </Pressable>
    );
};

type Styles = {
    xIcon: ImageStyle;
    xIconWrapper: ViewStyle;
};

const styles = StyleSheet.create<Styles>({
    xIconWrapper: {
        height: '100%',
        justifyContent: 'center',
    },
    xIcon: {
        width: 20,
        height: 20,
        zIndex: 1,
    },
});
