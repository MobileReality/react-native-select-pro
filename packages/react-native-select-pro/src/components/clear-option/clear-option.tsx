import React from 'react';
import type { ImageStyle, ViewStyle } from 'react-native';
import { Image, Pressable, StyleSheet } from 'react-native';

import { useClearOption } from './clear-option.hooks';

const iconSource = require('./../../assets/icons/x.png');

export const ClearOption = () => {
    const {
        disabled,
        clearOptionButtonProps,
        clearOptionImageProps,
        iconCustomStyles,
        containerCustomStyles,
        onPressClearOption,
    } = useClearOption();

    return (
        <Pressable
            accessibilityLabel="Clear a selected option"
            accessibilityRole="button"
            accessible={true}
            accessibilityState={{ disabled }}
            hitSlop={{ right: 3, left: 3 }}
            disabled={disabled}
            {...clearOptionButtonProps}
            style={[styles.xIconWrapper, containerCustomStyles]}
            onPress={onPressClearOption}
        >
            <Image
                source={iconSource}
                {...clearOptionImageProps}
                style={[styles.xIcon, iconCustomStyles]}
            />
        </Pressable>
    );
};

type Styles = {
    xIcon: ImageStyle;
    xIconWrapper: ViewStyle;
};

const styles = StyleSheet.create<Styles>({
    xIconWrapper: {
        height: '100%',
        justifyContent: 'center',
    },
    xIcon: {
        width: 20,
        height: 20,
        zIndex: 1,
    },
});
