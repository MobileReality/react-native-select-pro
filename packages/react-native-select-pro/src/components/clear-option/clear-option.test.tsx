import React from 'react';

import { ClearOption } from './clear-option';

describe('ClearOption', () => {
    it('should generate ClearOption snapshot', () => {
        const clearOption = (
            <ClearOption
                selectControlClearOptionA11yLabel={undefined}
                disabled={false}
                clearOptionStyles={{}}
                onPressRemove={() => {}}
            />
        );
        expect(clearOption).toMatchSnapshot();
    });
});
