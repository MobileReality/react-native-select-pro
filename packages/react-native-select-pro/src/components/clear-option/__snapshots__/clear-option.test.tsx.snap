// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClearOption should generate ClearOption snapshot 1`] = `
<View
  accessibilityLabel="Clear a chosen option"
  accessibilityRole="button"
  accessibilityState={
    {
      "disabled": false,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={
    {
      "left": 3,
      "right": 3,
    }
  }
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "height": "100%",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    source={
      {
        "testUri": "packages/react-native-select-pro/src/assets/icons/x.png",
      }
    }
    style={
      [
        {
          "height": 20,
          "width": 20,
          "zIndex": 1,
        },
        undefined,
      ]
    }
  />
</View>
`;
